### DISCUSSION TITLE
How to use with/has/preload in v5

### DISCUSSION THREAD

**Discussion by yarapolana** [adonisjs/core#1673] [General]
Hello there, I am converting my v4.1 project to v5 and I have some difficulties understanding and using the with, has, and preload relationships in v5.

Here is what I am trying to do and a reference to how I did these in v4.1
and my attempts in v5

<img src="https://user-images.githubusercontent.com/19730118/94366576-c5fbba00-00c8-11eb-8ace-cfddc837a846.png" width="400">

<img src="https://user-images.githubusercontent.com/19730118/94366848-42db6380-00ca-11eb-83b2-707c02fc7dd7.png" width="400">

<img src="https://user-images.githubusercontent.com/19730118/94367004-2c81d780-00cb-11eb-977c-3cd9ac8bdab7.png" width="500">

Tried accessing this page but it goes back to the same introduction page
https://preview.adonisjs.com/guides/database/query-builder 👈🏽 loop

<img width="894" alt="Screenshot 2020-09-27 at 13 20 51" src="https://user-images.githubusercontent.com/19730118/94367020-50451d80-00cb-11eb-8dbd-b2cb457e03a4.png">

**Comment by yarapolana** [comment]
After many failed attempts preloading like this worked

```
 const user = await auth.authenticate()
 await user.preload('profile')
 return user
```

wondering why this comes out empty
``` 
 const user = await auth.user.preload('profile')
```

**Reply by thetutlage** [reply]
Preloading means, you load the relationship from the database and set it on the existing instance. Which means

```ts
await user.preload('profile')

// will set following without returning anything
user.profile
```

Preloading from a single model instance is not that useful. However, preloading is useful when you are fetching more than one rows and want to load relationships at the same time.

```ts
const users = await User.query().preload('profile')

// here all users will have the `profile` property
users.forEach((user) => {
  user.profile
})
```

However, when working with a single instance (as you are), you can do following.

```ts
const profile = await auth.user!.related('profile').query().first()
```

To summarize. To keep the behavior of `preload` consistent it never returns any value but instead set the relationship on the parent model instance. Hope this helps :)

Regarding `with`. It has been replaced with `preload`
The `has` method works as it used to work earlier

**Reply by yarapolana** [reply]
Wow @thetutlage didn't know you worked on sundays. 
Thank you for the fast reply as usual.

For the `has` in the context of scopes it appears to only accept `undefined`

```
  public static hasRooms = scope(query => {
    return query.has('rooms') // <-- complains that "it is not assignable to parameter of type 'undefined'"
  })
```

My untested workaround was to use this

```
  public static hasRooms = scope(_ => {
    return Post.query().has('rooms')
  })
```

**Reply by thetutlage** [reply]
Yup. Since your questions are well constructed, there are usually easy to reply. 

The `scope` helper has no knowledge (that static level) that you were dealing with the query builder of the `Post` model. It is a typescript limitation. So you have to define the scope as follows:

```ts
public static hasRooms = scope<typeof Post>(query => {
   query.has('rooms')
})
```

**Reply by yarapolana** [reply]
Thank you a thousand. =) 
Enjoy the rest of the Sunday.
